Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,754 @@
+// -----------------------------------------------------------------------------
+///
+/// This file has the common implementation of Native protocols like HTTP,FTP,Data 
+/// These are libcurl helpers for the URLSession API code.
+/// - SeeAlso: https://curl.haxx.se/libcurl/c/
+/// - SeeAlso: NSURLSession.swift
+///
+// -----------------------------------------------------------------------------
+
+import CoreFoundation
+import Dispatch
+
+internal let enableLibcurlDebugOutput: Bool = {
+    return  (ProcessInfo.processInfo.environment["URLSessionDebugLibcurl"] != nil)
+}()
+internal let enableDebugOutput: Bool = {
+    return (ProcessInfo.processInfo.environment["URLSessionDebug"] != nil)
+}()
+
+class _NativeProtocol: URLProtocol, _EasyHandleDelegate {
+    internal var easyHandle: _EasyHandle!
+    internal var totalDownloaded = 0
+    internal lazy var tempFileURL: URL = {
+        let fileName = NSTemporaryDirectory() + NSUUID().uuidString + ".tmp"
+        _ = FileManager.default.createFile(atPath: fileName, contents: nil)
+        return URL(fileURLWithPath: fileName)
+    }()
+
+    public required init(task: URLSessionTask, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) {
+        self.internalState = _InternalState.initial
+        super.init(request: task.originalRequest!, cachedResponse: cachedResponse, client: client)
+        self.task = task
+        self.easyHandle = _EasyHandle(delegate: self)
+    }
+
+    public required init(request: URLRequest, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) {
+        self.internalState = _InternalState.initial
+        super.init(request: request, cachedResponse: cachedResponse, client: client)
+        self.easyHandle = _EasyHandle(delegate: self)
+    }
+
+    var internalState: _InternalState {
+        // We manage adding / removing the easy handle and pausing / unpausing
+        // here at a centralized place to make sure the internal state always
+        // matches up with the state of the easy handle being added and paused.
+        willSet {
+            if !internalState.isEasyHandlePaused && newValue.isEasyHandlePaused {
+                fatalError("Need to solve pausing receive.")
+            }
+            if internalState.isEasyHandleAddedToMultiHandle && !newValue.isEasyHandleAddedToMultiHandle {
+                task?.session.remove(handle: easyHandle)
+            }
+        }
+        didSet {
+            if !oldValue.isEasyHandleAddedToMultiHandle && internalState.isEasyHandleAddedToMultiHandle {
+                task?.session.add(handle: easyHandle)
+            }
+            if oldValue.isEasyHandlePaused && !internalState.isEasyHandlePaused {
+                fatalError("Need to solve pausing receive.")
+            }
+        }
+    }
+
+    func didReceive(data: Data) -> _EasyHandle._Action {
+        guard case .transferInProgress(var ts) = internalState else { fatalError("Received body data, but no transfer in progress.") }
```

Comment:
Can you insert a newline after the `{` and before the `fatalError`, along with `}` on its own line? That way the line size doesn't balloon for no reason.